metals/src/main/scala/scala/meta/internal/worksheets/MetalsWorksheetProvider.scala

```diff
@@ -0,0 +1,352 @@
+              for {
+                (binder, i) <- statement.binders.iterator.zipWithIndex
+                text <- Iterator[Iterator[Char]](
+                  if (isSingle) List.empty[Char].iterator
+                  else {
+                    val comma = if (i == 0) "" else ", "
+                    s"$comma${binder.name}=".iterator
+                  },
+                  pprint.PPrinter.BlackWhite
+                    .tokenize(
+                      binder.value,
+                      width = margin
+                    )
+                    .map(_.getChars)
+                    .filterNot(_.iterator.forall(_.isWhitespace))
+                    .flatMap(_.iterator)
+                    .filter {
+                      case '\n' => false
+                      case _ => true
+                    }
+                    .take(margin)
+                )
+                char <- text
+              } yield char
+            }
+          }
+        ).flatten.mkString
+        val hoverMessage: String = Iterator[Iterator[String]](
+          if (isEmptyValue) {
+            List.empty[String].iterator
+          } else {
+            for {
+              binder <- statement.binders.iterator
+              text <- Iterator(
+                "\n",
+                binder.name,
+                ": ",
+                binder.tpe.render(TPrintColors.BlackWhite),
+                " = "
+              ) ++ pprint.PPrinter.BlackWhite
+                .tokenize(binder.value, width = 100)
+                .map(_.plainText)
+            } yield text
+          },
+          statement.out.linesIterator.flatMap { line =>
+            Iterator("\n// ", line)
+          }
+        ).flatten.mkString
+        DecorationOptions(
+          range,
+          new l.MarkedString("scala", hoverMessage),
+          ThemableDecorationInstanceRenderOptions(
+            after = ThemableDecorationAttachmentRenderOptions(
+              contentText,
+              color = "green",
+              fontStyle = "italic"
+            )
+          )
+        )
+      }
+      languageClient.publishDiagnostics(
+        new PublishDiagnosticsParams(
+          path.toURI.toString(),
+          reporter.diagnostics.toSeq.asJava
+        )
+      )
+      decorations
+        .filterNot(_.renderOptions.after.contentText == commentHeader)
+        .toArray
+    }
+    decorations.getOrElse(Array.empty)
+  }
+
+  private def getContext(path: AbsolutePath): Option[Context] = {
+    for {
+      target <- buildTargets.inverseSources(path)
+      info <- buildTargets.scalaTarget(target)
+      scala <- info.info.asScalaBuildTarget
+      scalaVersion = scala.getScalaVersion
+      isSupported = ScalaVersions.isSupportedScalaVersion(scalaVersion)
+      _ = {
+        if (!isSupported) {
+          scribe.warn(
+            s"worksheet: unsupported Scala version '${scalaVersion}', to fix this problem use Scala version '${BuildInfo.scala212}' instead."
```

Comment:
do worksheets only work with Scala 2.12? If not, shouldn't we recommend 2.13 instead?

Reply:
Worksheets currently only work with 2.12.10. To support 2.11/2.13 I would prefer to move some of the logic here into mdoc and publish a stable `mdoc-interfaces` Java library. I felt it would be better to leave that for a separate PR